src/org/sosy_lab/common/NativeLibraries.java

```diff
@@ -105,55 +111,62 @@ public static Architecture guessVmArchitecture() {
-          case "32":
-            currentArch = Architecture.X86;
-            break;
-          case "64":
-            currentArch = Architecture.X86_64;
-            break;
-          default:
-            throw new UnsatisfiedLinkError(
-                "Unknown value for os.arch.data.model: '"
-                    + prop
-                    + "', please report this together with information about your system "
-                    + "(OS, architecture, JVM).");
-        }
-      } else {
+      osArch = Ascii.toLowerCase(osArch.replace(" ", ""));
+
+      switch (osArch) {
+        case "i386":
+        case "i686":
+        case "x86":
+          return Architecture.X86;
+        case "amd64":
+        case "x86_64":
+          return Architecture.X86_64;
+        case "arm":
+        case "armv6":
+        case "armv7":
+          return Architecture.ARM;
+        case "aarch64":
+          return Architecture.ARM64;
+        default:
+          throw new UnsatisfiedLinkError(
+              "Unknown value for os.arch: '"
+                  + StandardSystemProperty.OS_ARCH.value()
+                  + "'"
+                  + REPORT_MESSAGE);
+      }
+    }
 
-        prop = StandardSystemProperty.JAVA_VM_NAME.value();
-        if (!isNullOrEmpty(prop)) {
-          prop = Ascii.toLowerCase(prop);
-
-          if (prop.contains("32-bit") || prop.contains("32bit") || prop.contains("i386")) {
-
-            currentArch = Architecture.X86;
-          } else if (prop.contains("64-bit")
-              || prop.contains("64bit")
-              || prop.contains("x64")
-              || prop.contains("x86_64")
-              || prop.contains("amd64")) {
-
-            currentArch = Architecture.X86_64;
-          } else {
-            throw new UnsatisfiedLinkError(
-                "Unknown value for java.vm.name: '"
-                    + prop
-                    + "', please report this together with information about your system "
-                    + "(OS, architecture, JVM).");
-          }
-        } else {
-          throw new UnsatisfiedLinkError("Could not detect system architecture");
+    /**
+     * In most cases, the system architecture matches the JVM architecture. In some rare cases,
+     * architecture from 'os.arch' might not be sufficient. For example, we could use an emulator
+     * and execute a 32bit JVM on a 64bit system. We cannot check each possible combination and
+     * emulator, but check only the important ones.
+     */
+    private static Architecture downgradeArchIfRequired(Architecture arch) {
+      if (nullToEmpty(StandardSystemProperty.JAVA_VM_NAME.value()).matches("32-?bit")
+          || Objects.equals(System.getProperty("os.arch.data.model"), "32")
+          || Objects.equals(System.getProperty("sun.arch.data.model"), "32")) {
+        switch (arch) {
+          case X86_64:
+            return X86;
+          case ARM64:
+            return ARM;
+          default:
+            // $fall-through
```

Comment:
There is no fall-through into another case here, just remove it.

Reply:
done